[pt2e][quant] Make move_exported_model_to_train/eval idempotent

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142239

Summary: Before we would recompile the model unnecessarily even
if the model is already in the desired mode. For training
frameworks that assume `model.train()` is idempotent and calls
this before every single training step, this led to a bunch of
tiny graphs and poor performance. This commit makes these calls
no-ops if we're already in the target train/eval mode.

Test Plan:
python test/test_quantization -k TestQuantizePT2E.test_allow_exported_model_train_eval_idempotent

andrewor14 committed:
[pt2e][quant] Make move_exported_model_to_train/eval idempotent

Summary: Before we would recompile the model unnecessarily even
if the model is already in the desired mode. For training
frameworks that assume `model.train()` is idempotent and calls
this before every single training step, this led to a bunch of
tiny graphs and poor performance. This commit makes these calls
no-ops if we're already in the target train/eval mode.

Test Plan:
python test/test_quantization -k TestQuantizePT2E.test_allow_exported_model_train_eval_idempotent

[ghstack-poisoned]

andrewor14 committed:
Update on "[pt2e][quant] Make move_exported_model_to_train/eval idempotent"


Summary: Before we would recompile the model unnecessarily even
if the model is already in the desired mode. For training
frameworks that assume `model.train()` is idempotent and calls
this before every single training step, this led to a bunch of
tiny graphs and poor performance. This commit makes these calls
no-ops if we're already in the target train/eval mode.

Test Plan:
python test/test_quantization -k TestQuantizePT2E.test_allow_exported_model_train_eval_idempotent

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/export_utils.py:
```diff
@@ -195,7 +195,12 @@ def _move_exported_model_to_eval(model: torch.fx.GraphModule):
     """
+    if not model.training:
+        return model
+    model.training = False
```
is this attribute preserved in the transformations after export, will it be a problem if it's not?

andrewor14 replied:
Hmm good point, maybe safer to use a different private flag

andrewor14 committed:
Update on "[pt2e][quant] Make move_exported_model_to_train/eval idempotent"


Summary: Before we would recompile the model unnecessarily even
if the model is already in the desired mode. For training
frameworks that assume `model.train()` is idempotent and calls
this before every single training step, this led to a bunch of
tiny graphs and poor performance. This commit makes these calls
no-ops if we're already in the target train/eval mode.

Test Plan:
python test/test_quantization -k TestQuantizePT2E.test_allow_exported_model_train_eval_idempotent

[ghstack-poisoned]

andrewor14 committed:
Update on "[pt2e][quant] Make move_exported_model_to_train/eval idempotent"


Summary: Before we would recompile the model unnecessarily even
if the model is already in the desired mode. For training
frameworks that assume `model.train()` is idempotent and calls
this before every single training step, this led to a bunch of
tiny graphs and poor performance. This commit makes these calls
no-ops if we're already in the target train/eval mode.

Test Plan:
python test/test_quantization -k TestQuantizePT2E.test_allow_exported_model_train_eval_idempotent

[ghstack-poisoned]

andrewor14 commented:
@pytorchbot merge
